src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.netcoreapp.cs

```diff
@@ -20,5 +23,92 @@ public ReadOnlySpan<T> AsSpan(Range range)
             (int start, int length) = range.GetOffsetAndLength(self.Length);
             return new ReadOnlySpan<T>(self.array, start, length);
         }
+
+        /// <summary>
+        /// Removes the specified values from this list.
+        /// </summary>
+        /// <param name="items">The items to remove if matches are found in this list.</param>
+        /// <param name="equalityComparer">
+        /// The equality comparer to use in the search.
+        /// </param>
+        /// <returns>
+        /// A new list with the elements removed.
+        /// </returns>
+        public ImmutableArray<T> RemoveRange(ReadOnlySpan<T> items, IEqualityComparer<T>? equalityComparer = null)
+        {
+            var self = this;
+            self.ThrowNullRefIfNotInitialized();
+
+            if (items.IsEmpty || self.IsEmpty)
+            {
+                return self;
+            }
+
+            if (items.Length == 1)
+            {
+                return self.Remove(items[0], equalityComparer);
+            }
+
+#nullable disable
+            var itemsMultiSet = new Dictionary<T, int>(equalityComparer);
+#nullable restore
+            int nullValueCount = 0;
+            foreach (ref readonly T item in items)
+            {
+                if (item == null)
+                {
+                    nullValueCount++;
+                }
+                else
+                {
+#nullable disable
+                    ref int count = ref CollectionsMarshal.GetValueRefOrAddDefault(itemsMultiSet, item, out _);
+#nullable restore
+                    count++;
+                }
+            }
+
+            List<int>? indicesToRemove = null;
+            T[] selfArray = self.array!;
+            for (int i = 0; i < selfArray.Length; i++)
+            {
+                bool found = false;
+                if (selfArray[i] == null)
+                {
+                    if (nullValueCount > 0)
+                    {
+                        found = true;
+                        nullValueCount--;
+                    }
+                }
+                else
+                {
+#nullable disable
+                    ref int count = ref CollectionsMarshal.GetValueRefOrNullRef(itemsMultiSet, selfArray[i]);
+#nullable restore
+                    if (!Unsafe.IsNullRef(ref count))
```

Comment:
Likewise,
```C#
if (Unsafe.IsNullRef(ref count) || count-- == 0)
{
    continue;
}
```
I don't think there's a need to remove the entry once the count drops to zero.

Reply:
Purpose of removing the entry once found `count `will be zero is to avoid every further check in future count value check: so that for same value, only checking `Unsafe.IsNullRef(ref count)` will be enough.
Also I feel it is unnecessary to calculate `count` to negative value by `count--` in condition statement.

Although it uses more code, but seems during runtime, it runs shortly. Correct me if wrong, thanks.

Reply:
> Purpose of removing the entry once found count will be zero is to avoid every further check in future count value check

But it doesn't really avoid future lookups -- they simply return no result. On the other hand [removal is a nontrivial operation](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L968-L1034).

Reply:
Fixed, thanks.

> it doesn't really avoid future lookups

Previous thought is for this case (maybe not common) for example: 
input items: [1]
ImmutableArray: [1, 1, 1, 1, 1, 1, 1... ]
Then if we find `multiSet`has no more candidate for number 1 at first time and Remove it from `multiSet`, then next time to iterate further '1' in `ImmutableArray`, we will not need to check its `count`value again, because it cannot be fetched out from `multiSet`already.

But yes Removal is nontrivial operation. Thanks.